lib/theory/src/Theory/Constraint/Solver/ProofMethod.hs

```diff
@@ -485,31 +498,37 @@ rankGoals ctxt ranking tacticsList sys = case ranking of
       chooseError [] _ = error $ "No tactic has been written in the theory file"
       chooseError _  t = error $ "The tactic specified ( "++(show $ _name t)++" ) is not written in the theory file, please chose among the following: "++(show definedHeuristic)
 
+isFinished :: ProofContext -> System -> Maybe Result
+isFinished ctxt sys
+  | isInitialSystem sys = Nothing
+  | not $ null cs = Just $ Contradictory (Just $ head cs)
+  | null ogs && stFinished = Just Solved
+  | null ogs && not stFinished = Just Unfinishable
+  | otherwise = Nothing
+  where
+    cs = contradictions ctxt sys
+    ogs = openGoals sys
+    stFinished = finishedSubterms ctxt sys
+
 -- | Use a 'GoalRanking' to generate the ranked, list of possible
 -- 'ProofMethod's and their corresponding results in this 'ProofContext' and
--- for this 'System'. If the resulting list is empty, then the constraint
--- system is solved.
+-- for this 'System'.
 rankProofMethods :: GoalRanking ProofContext -> [Tactic ProofContext] -> ProofContext -> System
                  -> [(ProofMethod, (M.Map CaseName System, String))]
 rankProofMethods ranking tactics ctxt sys =
-  let rankedGoals = rankGoals ctxt ranking tactics sys $ openGoals sys
-      cs = contradictions ctxt sys
-  in case rankedGoals of
-    Ranking gs (Just ApplySorry)
-      | null cs && not (null gs) -> [(Sorry (Just "Oracle ranked no goals"), (M.empty, ""))]
-    Ranking gs _ -> do
-      (m, expl) <-
-              (contradiction <$> cs)
-          <|> (case L.get pcUseInduction ctxt of
-                AvoidInduction -> [(Simplify, ""), (Induction, "")]
-                UseInduction   -> [(Induction, ""), (Simplify, "")]
-              )
-          <|> (solveGoalMethod <$> gs)
-      case execProofMethod ctxt m sys of
-        Just cases -> return (m, (cases, expl))
-        Nothing    -> []
+  let Ranking (map solveGoalMethod -> goals) instr = rankGoals ctxt ranking tactics sys (openGoals sys)
+      initialMethods = case L.get pcUseInduction ctxt of
+        AvoidInduction -> [(Simplify, ""), (Induction, "")]
+        UseInduction   -> [(Induction, ""), (Simplify, "")]
+      proofMethods = if isInitialSystem sys then initialMethods ++ goals else goals
+      stoppingMethod =    (Finished <$> isFinished ctxt sys)
+                      <|> (Sorry (Just "Oracle ranked no goals") <$ instr)
+  in execMethods $ maybe proofMethods ((:[]) . (,"")) stoppingMethod
   where
-    contradiction c                    = (Contradiction (Just c), "")
+    execMethods = mapMaybe execMethod
+    execMethod (m, expl) = do
+      cases <- execProofMethod ctxt m sys
```

Comment:
Shouldn't this be a call to `checkAndExecProofMethods` now that `execProofMethod` no longer checks whether the method is valid? Or do we rely on `execProofMethod` returning `Nothing` if the execution fails?

Reply:
`rankProofMethods` computes all applicable methods itself, i.e., they must be valid (or the computation is buggy, but this shouldn't be fixed by checking that they are valid imo). Thus, checking whether it's valid is not necessary. At this point in the code, it cannot come from an "external" source such as a loaded proof file.